NXP-29984 Runtime Registries Migrations #1

atchertchian opened the pull request.

nuxeojenkins commented:
View issues in JIRA:
- [NXP-29984](https://jira.nuxeo.com/browse/NXP-29984): Migrate Runtime Contributions to XRegistry Annotations
- [NXP-29884](https://jira.nuxeo.com/browse/NXP-29884): Setup build and release Slack notifications on Nuxeo JSF UI

efge commented on modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java:
```diff
@@ -94,7 +98,15 @@ protected void register(Context ctx, XAnnotatedObject xObject, Element element)
-            contrib = xObject.newInstance(ctx, element, contributions.get(id));
+            Object contribution = contributions.get(id);
+            if (contribution != null && xObject.getCompatWarnOnMerge() && !merge.hasValue(ctx, element)) {
+                log.warn(String.format(
+                        "The contribution with id '%s' has been implicitely merged: the compatibility mechanism on its"
+                                + " descriptor class '%s' detected it, and"
+                                + " the attribute merge=\"true\" should be added to this definition.",
+                        id, contribution.getClass().getName()));
```
You should use `{}` parameters instead of `String.format` in logs. Same in `SingleRegistry`.

efge commented on modules/runtime/nuxeo-runtime-test/src/test/java/org/nuxeo/runtime/registry/SampleWarnOnMergeDescriptor.java:
```diff
@@ -0,0 +1,22 @@
+package org.nuxeo.runtime.registry;
```
Missing copyright header.

efge commented on modules/core/nuxeo-core-event/src/main/java/org/nuxeo/ecm/core/event/impl/EventServiceImpl.java:
```diff
@@ -117,33 +124,35 @@ void push(Event event) {
-        EventDispatcherDescriptor dispatcherDescriptor = dispatchers.getDispatcherDescriptor();
-        if (dispatcherDescriptor != null) {
-            List<EventPipeDescriptor> pipes = registeredPipes.getPipes();
+        listenerDescriptors = this.<EventListenerList> getRegistry(EVENT_LISTENER_XP);
```
Are you sure the generic is needed here, given that there's a typed target assignment?

atchertchian replied:
No indeed :+1:

efge commented on modules/core/nuxeo-core-event/src/main/java/org/nuxeo/ecm/core/event/impl/EventServiceImpl.java:
```diff
@@ -117,33 +124,35 @@ void push(Event event) {
-        if (dispatcherDescriptor != null) {
-            List<EventPipeDescriptor> pipes = registeredPipes.getPipes();
+        listenerDescriptors = this.<EventListenerList> getRegistry(EVENT_LISTENER_XP);
+        Optional<EventDispatcherDescriptor> optDispatcher = this.<SingleRegistry> getRegistry(EVENT_DISPATCHER_XP)
+                                                                .getContribution();
+        if (optDispatcher.isPresent()) {
```
If there's no checked exception messing things up, this `if` block would be a lambda of `ifPresent` on the previous call.

efge reviewed (changes requested):
Details

efge commented on modules/platform/nuxeo-platform-audit/nuxeo-platform-audit-core/src/main/java/org/nuxeo/ecm/platform/audit/service/extension/AuditBackendDescriptor.java:
```diff
@@ -38,14 +42,18 @@
-        }
-        return ((DefaultComponent)Framework.getRuntime().getComponent(requiredComponent)).getApplicationStartedOrder()+1;
+        return requires.stream()
+                       .map(r -> Framework.getRuntime().getComponent(r))
```
`.map(Framework.getRuntime()::getComponent)` probably works here

efge commented on modules/platform/nuxeo-platform-audit/nuxeo-platform-audit-core/src/main/java/org/nuxeo/ecm/platform/audit/service/extension/AuditBackendDescriptor.java:
```diff
@@ -38,14 +42,18 @@
+        return requires.stream()
+                       .map(r -> Framework.getRuntime().getComponent(r))
+                       .map(DefaultComponent.class::cast)
+                       .map(DefaultComponent::getApplicationStartedOrder)
+                       .mapToInt(v -> v)
```
Just `.mapToInt(DefaultComponent::getApplicationStartedOrder)`

efge commented on modules/platform/nuxeo-platform-audit/nuxeo-platform-audit-core/src/main/java/org/nuxeo/ecm/platform/audit/service/NXAuditEventsService.java:
```diff
@@ -56,6 +56,8 @@
  */
 public class NXAuditEventsService extends DefaultComponent implements ComponentManager.Listener {
 
+    protected static final Logger log = LogManager.getLogger(NXAuditEventsService.class);
```
Can you make it `private` while we're at it?

efge commented on modules/platform/nuxeo-binary-metadata/src/main/java/org/nuxeo/binary/metadata/internals/MetadataRuleDescriptor.java:
```diff
@@ -108,4 +114,13 @@ public boolean equals(Object obj) {
 
+    @Override
+    public int compareTo(MetadataRuleDescriptor o) {
+        int order = getOrder().compareTo(o.getOrder());
+        if (order != 0) {
+            return order;
+        }
+        return getId().compareTo(o.getId());
```
Could be

    return Comparator.comparing(MetadataRuleDescriptor::getOrder)
                     .thenComparing(MetadataRuleDescriptor::getId)
                     .compare(this, o);

efge reviewed (changes requested):
Stream details

efge commented on modules/platform/nuxeo-binary-metadata/src/main/java/org/nuxeo/binary/metadata/internals/MetadataRuleDescriptor.java:
```diff
@@ -108,4 +115,11 @@ public boolean equals(Object obj) {
 
+    @Override
+    public int compareTo(MetadataRuleDescriptor o) {
+        return Comparator.comparing(MetadataRuleDescriptor::getOrder)
+                         .thenComparing(MetadataRuleDescriptor::getId)
+                         .compare(this, o);
```
Actually we could even put `Comparator.comparing(MetadataRuleDescriptor::getOrder).thenComparing(MetadataRuleDescriptor::getId)` in a constant but not worth it here given the rare usage

atchertchian replied:
Actually rules are sorted on the fly when checking a filter, so maybe it's worth optimizing this?

efge replied:
Ah yes maybe then

kevinleturc reviewed (commented):
Nice work 👍 , some details.

atchertchian commented on modules/core/nuxeo-core-mimetype/src/main/java/org/nuxeo/ecm/platform/mimetype/service/MimetypeDescriptorRegistry.java:
```diff
@@ -0,0 +1,83 @@
+
+    private static final Logger log = LogManager.getLogger(MimetypeDescriptorRegistry.class);
+
+    protected Map<String, MimetypeEntry> mimetypeByNormalizedRegistry = Collections.synchronizedMap(new HashMap<>());
+
+    protected Map<String, MimetypeEntry> mimetypeByExtensionRegistry = Collections.synchronizedMap(new HashMap<>());
```
Not really, I just used the same than when I needed a LinkedHashMap, here it's not needed indeed.

atchertchian replied:
Fixed all unneeded occurrences :+1:

atchertchian commented:
Addressed feedback + adjusted changes to AuditBackendDescriptor to address test issues (some required components are sometimes null in tests), see https://github.com/nuxeo/nuxeo/pull/4575/commits/af9020be54ebe862c97dd1ddcbdf8c4d368bfc90

Also fixed some commit messages (wrong jira issue reference)

atchertchian commented:
Also realized that test failures on AuditBackendDescriptor were linked to unwanted merges on this descriptor --> disabled merge on this registry.

atchertchian commented:
(last force push was just to address a javadoc issue on actions)

atchertchian commented:
There are test failures linked to ConcurrentHashMap replacement usage (NPE when calling entry with a null key)

atchertchian commented:
Made code safer (impacting LifeCycleRegistry, MimetypeDescriptorRegistry, NotificationRegistryImpl, ActionRegistry)
